File: src/main/java/fi/quanfoxes/parser/nodes/LinkNode.java
package fi.quanfoxes.parser.nodes;

import fi.quanfoxes.Types;
import fi.quanfoxes.lexer.Operators;
import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Contextable;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Resolvable;

public class LinkNode extends OperatorNode implements Resolvable, Contextable {

    public LinkNode() {
        super(Operators.DOT);
    }

    private Context getContext(Node node) throws Exception {
        if (node instanceof Contextable) {
            Contextable contextable = (Contextable)node;
            return contextable.getContext();
        }

        return null;
    }

    @Override
    public Node resolve(Context base) throws Exception {
        Node left = first();
        Node right = last();

        if (left instanceof Resolvable) {
            Resolvable resolvable = (Resolvable)left;
            Node resolved = resolvable.resolve(base);

            if (resolved != null) {
                left.replace(resolved);
                left = resolved;
            }
        }

        if (right instanceof Resolvable) {
            Context context = getContext(left);

            if (context == Types.UNKNOWN) {
                throw new Exception("Couldn't resolve the type of the left hand side");
            }

            Node resolved;

            if (right instanceof UnresolvedFunction) {
                UnresolvedFunction function = (UnresolvedFunction)right;
                resolved = function.solve(base, context);
            }
            else {
                Resolvable resolvable = (Resolvable)right;
                resolved = resolvable.resolve(context);
            }

            if (resolved != null) {
                right.replace(resolved);
                right = resolved;
            }
        }

        return null;
    }

    @Override
    public Context getContext() throws Exception {
        Node right = last();
        return getContext(right);
    }
}

File: Zigzag/Parser/Nodes/LinkNode.cs
public class LinkNode : OperatorNode, IResolvable, IType
{
	public LinkNode(Node left, Node right) : base(Operators.DOT)
	{
		SetOperands(left, right);
	}

	private Type? GetNodeContext(Node node)
	{
		if (node is IType type)
		{
			return type.GetType();
		}

		return null;
	}

	public Node? Resolve(Context environment)
	{
		if (Left is IResolvable a)
		{
			var resolved = a.Resolve(environment);

			if (resolved != null)
			{
				Left.Replace(resolved);
				First = resolved;
			}
		}

		if (Right is IResolvable b)
		{
			var context = GetNodeContext(Left);

			if (context == Types.UNKNOWN)
			{
				return null;
			}

			Node? resolved;

			if (Right.GetNodeType() == NodeType.UNRESOLVED_FUNCTION)
			{
				var function = (UnresolvedFunction)Right;
				resolved = function.Solve(environment, context);
			}
			else
			{
				resolved = b.Resolve(context);
			}

			if (resolved != null)
			{
				Right.Replace(resolved);
				Last = resolved;
			}
		}

		return null;
	}

	public override Type? GetType()
	{
		return GetNodeContext(Right);
	}

	public override NodeType GetNodeType()
	{
		return NodeType.LINK_NODE;
	}

	public Status GetStatus()
	{
		return Status.OK;
	}
}